File: Dynamic-Programming/44. Wildcard Matching/wildcardMatching.py
class Solution:
    def isMatch(self, s: str, p: str) -> bool:
        dp = [[False for j in range(len(p) + 1)] for i in range(len(s) + 1)]
        
        # [0][0] is "" vs ""
        dp[0][0] = True
        # for the first row, set values to true until we see the first non-* character
        for j in range(1, len(p) + 1):
            if p[j - 1] != "*":
                break
            dp[0][j] = True
        
        for i in range(1, len(s) + 1):
            for j in range(1, len(p) + 1):
                if p[j - 1] in {s[i - 1], "?"}:
                    dp[i][j] = dp[i - 1][j - 1]
                elif p[j - 1] == "*":
                    dp[i][j] = dp[i - 1][j] or dp[i][j - 1]
                # 3rd case is p[j - 1] != s[i - 1], return False
                # but we already set it to false
        
        return dp[-1][-1]

File: Dynamic-Programming/44. Wildcard Matching/wildcardMatching.js
/**
 * @param {string} s
 * @param {string} p
 * @return {boolean}
 */
var isMatch = function (s, p) {
    let n = s.length;
    let m = p.length;

    let dp = new Array(n + 1);
    for (let i = 0; i < dp.length; i++) {
        dp[i] = new Array(m + 1).fill(false)
    }
    dp[0][0] = true;
    for (let i = 1; i < m + 1; i++) {
        if (p[i - 1] !== "*") break;
        dp[0][i] = true
    }

    for (let i = 1; i < n + 1; i++) {
        for (let j = 1; j < m + 1; j++) {
            if (p[j - 1] === "?" || p[j - 1] === s[i - 1]) {
                dp[i][j] = dp[i - 1][j - 1];
            } else if (p[j - 1] === "*") {
                dp[i][j] = dp[i][j - 1] || dp[i - 1][j];
            }
        }
    }

    return dp[n][m];
};